grader_systems/src/stage/valgrind.rs

```diff
@@ -0,0 +1,1123 @@
+        }
+
+        // replace all absolute paths with basename
+        let re = Regex::new(r"(\W|^)(?:\/[^\/\s]+)+\/([^\/\s]+)\b")?;
+        let repl = re.replace_all(&contents, "$1$2");
+
+        Ok(repl.to_string())
+    }
+}
+
+#[async_trait]
+impl<E: ProcessExecutor> Executor for Valgrind<E> {
+    type Output = StageResult;
+
+    async fn run(&self, ws: &Path) -> Result<Self::Output> {
+        let mut sect = Section::new("Valgrind");
+        let mut results_sect = Section::new("Output:");
+        let mut run_updates = StatusUpdates::default();
+        debug!("running program");
+
+        let executable = ws.join(&self.executable);
+        if !executable.exists() {
+            return Err(anyhow!(
+                "Could not find student executable at {:?}",
+                executable.display()
+            ));
+        }
+
+        if let Some(v) = &self.config.suppressions {
+            for supp in v {
+                if !ws.join(&supp).exists() {
+                    return Err(anyhow!("Could not find valgrind suppressions at {}", supp));
+                }
+            }
+        }
+
+        let cmd = self.gen_cmd(ws);
+        sect.add_content(("Run Command", format!("{}", cmd).code()));
+
+        let res = cmd.run_with(&self.executor).await?;
+
+        let log = self.read_logfile(ws)?;
+
+        if !res.status.completed() {
+            match res.status {
+                ExitStatus::Timeout(to) => {
+                    run_updates.add_update(
+                        Update::new_fail("running valgrind", PointQuantity::FullPoints).notes(
+                            format!(
+                                "Your submission timed out after {} seconds :(",
+                                to.as_secs()
+                            ),
+                        ),
+                    );
+                    sect.add_content(run_updates);
+                }
+                ExitStatus::Signal(sig) => match sig {
+                    SignalType::SegFault => {
+                        run_updates.add_update(
+                            Update::new_fail("running valgrind", PointQuantity::FullPoints)
+                                .notes("Your submission was killed by SIGSEGV!"),
+                        );
+                        results_sect.add_content(log);
+                        sect.add_content(run_updates);
+                        sect.add_content(Content::SubSection(results_sect));
+                    }
+                    SignalType::Abort => {
+                        run_updates.add_update(
+                            Update::new_fail("running valgrind", PointQuantity::FullPoints)
+                                .notes("Your submission was killed by SIGABRT!"),
+                        );
+                        results_sect.add_content(log);
+                        sect.add_content(run_updates);
+                        sect.add_content(Content::SubSection(results_sect));
+                    }
```

Comment:
If the submission aborts or gets a segfault, how does valgrind report that? Will it return the the exit codes expected with those failure types? Just wondering if valgrind will catch these somehow and return a different error code like 2 or something.

Reply:
Valgrind normally returns 128+signal on linux if it fails with SIGSEGV or SIGABRT, regardless of the setting of --error-exitcode. I might need to do some further tests to see if ExitStatus is guaranteed to be Signal if this happens (assuming SIGABRT and SIGSEGV are the only two signals possible in this case)

Reply:
process will check exit code to determine if the signal was abort or segv (6 and 11 respectively). Let's double check that valgrind propagates those errors up. You can use the test program I have in genos/resources to do that.